algo/src/main/java/org/neo4j/graphalgo/impl/MSTPrim.java

```diff
@@ -28,63 +33,95 @@ public MSTPrim(Graph graph) {
      * @param startNode the node to start the evaluation from
      * @return a container of the transitions in the minimum spanning tree
      */
-    public MST compute(int startNode) {
+    public MinimumSpanningTree compute(int startNode) {
 
         final LongMinPriorityQueue queue = new LongMinPriorityQueue();
-        final MST mst = new MST(graph.nodeCount());
-        final BitSet visited = new BitSet(graph.nodeCount());
+        final MinimumSpanningTree mst = new MinimumSpanningTree();
+        final BitSet visited = new BitSet(idMapping.nodeCount());
 
         // initially add all relations from startNode to the priority queue
         visited.set(startNode);
-        graph.forEachRelationship(startNode, Direction.BOTH, (sourceNodeId, targetNodeId, relationId, weight) -> {
+        iterator.forEachRelationship(startNode, Direction.BOTH, (sourceNodeId, targetNodeId, relationId, weight) -> {
             queue.add(combineIntInt(startNode, targetNodeId), weight);
         });
         while (!queue.isEmpty()) {
             // retrieve cheapest transition
+            final double cost = queue.topCost();
             final long transition = queue.pop();
             final int nodeId = getTail(transition);
             if (visited.get(nodeId)) {
                 continue;
             }
             visited.set(nodeId);
             // add to mst
-            mst.add(transition);
+            mst.add(getHead(transition), getTail(transition), cost);
             // add new candidates
-            graph.forEachRelationship(nodeId, Direction.BOTH, (sourceNodeId, targetNodeId, relationId, weight) -> {
+            iterator.forEachRelationship(nodeId, Direction.BOTH, (sourceNodeId, targetNodeId, relationId, weight) -> {
                 queue.add(combineIntInt(nodeId, targetNodeId), weight);
             });
         }
 
         return mst;
     }
 
+    public static class MinimumSpanningTree implements WeightedAllRelationshipIterator {
 
-    public static class MST {
+        private final LongDoubleMap costMap;
 
-        private final long[] data;
-        private int offset = 0;
+        public MinimumSpanningTree() {
+            this(8);
+        }
 
-        public MST(int capacity) {
-            data = new long[capacity];
+        public MinimumSpanningTree(int expectedElements) {
+            costMap = new LongDoubleScatterMap(expectedElements);
         }
 
-        /**
-         * get the element count in this mst
-         */
-        public int size() {
-            return offset;
+        public void add(int sourceNodeId, int targetNodeId, double cost) {
```

Comment:
We split the value into ints in line 57 and combine them again in 80. Change method signature (combinedSourceTarget:long, cost:double):void